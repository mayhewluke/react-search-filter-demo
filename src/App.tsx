import React from "react";

import Filter from "./Filter";
import SearchInput from "./SearchInput";

const styles = {
  headerBar: {
    backgroundColor: "#0af",
    display: "flex",
    height: "35px",
    justifyContent: "flex-end",
    padding: "5px",
  },
};

class App extends React.Component<{}, { query: string }> {
  constructor(props: {}) {
    super(props);
    this.state = {
      query: "",
    };
  }

  public onQueryChange = (query: string) => {
    this.setState({ query });
  };

  public render() {
    const { query } = this.state;
    return (
      <div>
        <div style={styles.headerBar}>
          <SearchInput onChange={this.onQueryChange} />
        </div>
        <div>
          <Filter query={query} />
        </div>
      </div>
    );
  }
}

export default App;
