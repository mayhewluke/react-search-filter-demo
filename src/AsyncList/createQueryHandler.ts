import { merge, Observable, of, Subject } from "rxjs";
import { ajax } from "rxjs/ajax";
import {
  catchError,
  debounceTime,
  map,
  share,
  startWith,
  switchMap,
  withLatestFrom,
} from "rxjs/operators";

import {
  AsyncValue,
  AsyncValueTag,
  error,
  init,
  loading,
  success,
} from "common/async";

// Convenience function for pulling results out of an AsyncValue, or returning
// null if it was of a type that has no value (or if the results were null).
const valueOrNull = <T>(asyncValue: AsyncValue<T>) => {
  switch (asyncValue.state) {
    case AsyncValueTag.LOADING:
    case AsyncValueTag.SUCCESS:
      return asyncValue.results;
    default:
      return null;
  }
};

type Transform<T> = (data: any) => T;

const fetchResults = <T>(
  url: string,
  transform: Transform<T>,
): Observable<AsyncValue<any>> =>
  ajax.getJSON(url).pipe(
    map((response: any) => success(transform(response))),
    catchError(() => of(error<any>())),
  );

const createQueryHandler = <T>(transform: Transform<T>) => {
  const subject = new Subject<string>();

  // results$ needs to be shared or else it will be a cold observable - which
  // means that when loading$ subscribes to it (in addition to it being
  // subscribed to when merged into the output stream) it will run through its
  // pipeline again, causing another ajax request
  const results$ = subject.pipe(
    debounceTime(500), // TODO make time length configurable
    switchMap(url => fetchResults(url, transform)),
    share(),
  );

  const loading$ = subject.pipe(
    withLatestFrom(results$.pipe(startWith(init()))),
    map(([_, results]) => loading(valueOrNull(results))),
  );

  return {
    handler: subject.next.bind(subject),
    results$: merge(loading$, results$),
  };
};

export default createQueryHandler;
