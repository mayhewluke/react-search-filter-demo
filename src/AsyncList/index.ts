export {
  default as AsyncList,
  Meal,
  Props as AsyncListProps,
} from "./AsyncList";
export { default as createQueryHandler } from "./createQueryHandler";
