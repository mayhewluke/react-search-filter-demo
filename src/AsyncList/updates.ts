import { Meal } from "./reducer";

export enum UpdateTag {
  SEARCH_START = "SEARCH_START",
  FETCH_SUCCESS = "FETCH_SUCCESS",
  FETCH_FAIL = "FETCH_FAIL",
  RESET = "RESET",
}

export type Update =
  | { type: UpdateTag.SEARCH_START }
  | { type: UpdateTag.FETCH_SUCCESS; payload: Meal[] }
  | { type: UpdateTag.FETCH_FAIL }
  | { type: UpdateTag.RESET };

export const searchStart = (): Update => ({
  type: UpdateTag.SEARCH_START,
});

export const fetchSuccess = (meals: Meal[]): Update => ({
  payload: meals,
  type: UpdateTag.FETCH_SUCCESS,
});

export const fetchFail = (): Update => ({ type: UpdateTag.FETCH_FAIL });

export const reset = (): Update => ({ type: UpdateTag.RESET });
