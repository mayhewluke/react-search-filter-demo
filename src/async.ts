export enum AsyncValueTag {
  INIT = "INIT",
  LOADING = "LOADING",
  SUCCESS = "SUCCESS",
  ERROR = "ERROR",
}

export type AsyncValue<T> =
  | { state: AsyncValueTag.INIT }
  | { state: AsyncValueTag.LOADING; results: T | null }
  | { state: AsyncValueTag.SUCCESS; results: T }
  // Normally we would also want to have some kind of payload here, such as
  // `error: string`, but since we don't use that in this demo we omit it here.
  | { state: AsyncValueTag.ERROR };

// Convenience functions for constructing the above types, since TypeScript
// doesn't make working with tagged unions as easy as say, Haskell or Elm
export const init = <T>(): AsyncValue<T> => ({ state: AsyncValueTag.INIT });
export const loading = <T>(previousResults: T | null): AsyncValue<T> => ({
  results: previousResults,
  state: AsyncValueTag.LOADING,
});
export const success = <T>(results: T): AsyncValue<T> => ({
  results,
  state: AsyncValueTag.SUCCESS,
});
export const error = <T>(): AsyncValue<T> => ({
  state: AsyncValueTag.ERROR,
});
