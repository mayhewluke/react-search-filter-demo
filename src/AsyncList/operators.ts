import { Observable, of } from "rxjs";
import { ajax } from "rxjs/ajax";
import { catchError, distinctUntilChanged, map, pluck } from "rxjs/operators";

import { AsyncValue } from "async";

import { fetchFail, fetchSuccess, Update } from "./updates";

export interface Props {
  query: string;
  render: React.SFC<Meal>;
}

export type State = AsyncValue<Meal[]>;

export interface Meal {
  idMeal: string; // id
  strMeal: string; // Name
}

const url = (query: string) =>
  `https://www.themealdb.com/api/json/v1/1/search.php?s=${query}`;

export const fetchResults = (query: string): Observable<Update> =>
  ajax.getJSON(url(query)).pipe(
    map((response: any) => fetchSuccess(response.meals || [])),
    catchError(() => of(fetchFail())),
  );

// Can be used as a `pipe` operator
export const whenQueryChanges = (
  props$: Observable<Props>,
): Observable<string> =>
  props$.pipe(
    pluck<Props, Props["query"]>("query"), // extract the `query` property
    distinctUntilChanged(), // filter out consecutive duplicates
  );
