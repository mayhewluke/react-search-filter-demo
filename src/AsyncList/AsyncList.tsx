import React from "react";
import { FaSpinner } from "react-icons/fa";
import { Observable } from "rxjs";
import { map } from "rxjs/operators";

import { AsyncValue, caseOf } from "common/async";
import { componentFromStream } from "common/streamHelpers";

export interface Props {
  results: AsyncValue<Meal[]>;
  render: React.SFC<Meal>;
  errorComponent: JSX.Element;
  noResultsComponent: JSX.Element;
}

export interface Meal {
  idMeal: string; // id
  strMeal: string; // Name
}

const renderItems = (items: Meal[] | null, render: React.SFC<Meal>) => {
  if (items === null) {
    return null;
  }
  return items.map(render);
};

const AsyncList = componentFromStream((props$: Observable<Props>) => {
  return props$.pipe(
    map(({ results, render, errorComponent, noResultsComponent }) =>
      caseOf(results, {
        // TODO pass the error in so errorComponent can provide useful contextual info
        error: () => errorComponent,
        init: () => null,
        loading: items => (
          <div>
            <FaSpinner data-testid="loading" />
            {renderItems(items, render)}
          </div>
        ),
        success: items =>
          items.length > 0 ? renderItems(items, render) : noResultsComponent,
      }),
    ),
  );
});

export default AsyncList;
