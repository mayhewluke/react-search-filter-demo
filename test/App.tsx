jest.mock("AsyncList", () => ({
  AsyncList: jest.fn(),
  createQueryHandler: jest.fn(),
}));

import React from "react";
import { fireEvent, render } from "react-testing-library";
import { Subject } from "rxjs";

import App from "App";
import { AsyncList, AsyncListProps, createQueryHandler } from "AsyncList";
import { init, success } from "common/async";

describe("App component", () => {
  const listProps = (key: keyof AsyncListProps) => {
    const calls = (AsyncList as any).mock.calls;
    return calls[calls.length - 1][0][key];
  };
  const handler = jest.fn();
  let results$: Subject<any>;
  beforeEach(() => {
    (AsyncList as any).mockImplementation(() => "AsyncList");
    results$ = new Subject();
    (createQueryHandler as any).mockImplementation(() => ({
      handler,
      results$,
    }));
  });

  it("sets AsyncList's results to the init state", () => {
    render(<App />);
    expect(listProps("results")).toEqual(init());
  });

  describe("when the input value is changed", () => {
    it("updates the noResultsComponent on AsyncList", () => {
      const input = render(<App />).container.querySelector("input")!;

      fireEvent.change(input, { target: { value: "foo" } });

      expect(listProps("noResultsComponent")).toMatchSnapshot();
    });

    it("initiates a request for the given query string", () => {
      const input = render(<App />).container.querySelector("input")!;
      const query = "foo";
      const url = `https://www.themealdb.com/api/json/v1/1/search.php?s=${query}`;

      fireEvent.change(input, { target: { value: "foo" } });

      expect(handler).toHaveBeenCalledTimes(1);
      expect(handler).toHaveBeenCalledWith(url);
    });

    describe("when createQueryHandler emits", () => {
      const results = success(["chicken"]);

      it("passes the extracted results to AsyncList", () => {
        render(<App />);

        expect(listProps("results")).not.toEqual(results);
        results$.next(results);

        expect(listProps("results")).toEqual(results);
      });
    });

    describe("when the value is cleared", () => {
      it("resets the results for AsyncList back to the init state", () => {
        const expected = init();
        const input = render(<App />).container.querySelector("input")!;

        // Must ensure that the input has text, or else setting it to empty
        // doesn't result in a change
        fireEvent.change(input, { target: { value: "bar" } });
        // Must also ensure that the results have been changed, since we've
        // stubbed out the mechanism that would do so based on query change
        results$.next(success([]));
        expect(listProps("results")).not.toEqual(expected);

        fireEvent.change(input, { target: { value: "" } });

        expect(listProps("results")).toEqual(expected);
      });
    });
  });

  // Ideally we'd test that the render and errorComponent on AsyncList are set
  // properly, and maybe use snapshot testing to check the overall
  // layout/styling, but that's currently omitted in the interest of time
});
