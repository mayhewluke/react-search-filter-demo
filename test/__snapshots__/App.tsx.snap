// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component when the input value is changed updates the noResultsComponent on AsyncList 1`] = `
<span>
  No results found for "
  foo
  "
</span>
`;
